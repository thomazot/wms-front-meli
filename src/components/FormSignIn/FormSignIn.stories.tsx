import { Story, Meta } from '@storybook/react'
import FormSignIn, { FormSignInProps } from './FormSignIn'

export default {
  title: 'Form/FormSignIn',
  component: FormSignIn,
  argTypes: {
    onSubmit: { action: 'submit' }
  }
} as Meta

export const Default: Story<FormSignInProps> = (args) => (
  <div style={{ maxWidth: 400, margin: 'auto' }}>
    <FormSignIn {...args} />
  </div>
)
