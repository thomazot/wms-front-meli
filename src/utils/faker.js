import faker from 'faker'

export const generator = (schema, min = 1, max) => {
  max = max || min
  return Array.from({
    length: faker.datatype.number({ min: min, max: max })
  }).map(() =>
    Object.keys(schema).reduce((entity, key) => {
      entity[key] = String(schema[key]).match(/^{{/)
        ? faker.fake(schema[key])
        : schema[key]
      return entity
    }, {})
  )
}

const productsSchema = {
  id: '{{datatype.uuid}}',
  title: '{{company.companyName}} {{company.companySuffix}}',
  price: '{{commerce.price}}',
  thumbnail: '/images/img.png',
  permalink: 'https://www.mercadolivre.com.br/produto/123',
  rating: 4,
  currency_id: 'BRL'
}

export const productsData = generator(productsSchema, 9, 9)

const menuSchema = {
  id: '{{datatype.uuid}}',
  name: '{{company.companyName}} {{company.companySuffix}}'
}

export const menuData = generator(menuSchema, 10, 20)
