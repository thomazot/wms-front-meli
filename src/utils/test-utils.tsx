import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'
import theme from 'styles/theme'

import { WishlistProvider } from 'hooks/use-wishlist'

type CustomRenderProps = Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  { ...renderOptions }: CustomRenderProps = {}
) =>
  render(
    <WishlistProvider>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </WishlistProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
