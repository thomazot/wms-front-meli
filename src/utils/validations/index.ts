import Joi from 'joi'

const fieldsValidations = {
  email: Joi.string()
    .email({ tlds: { allow: false } })
    .required(),
  password: Joi.string().required()
}

function getFieldErros(objError: Joi.ValidationResult) {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((detail) => {
      errors[detail.path.join('.')] = detail.message
    })
  }

  return errors
}

export type FieldErrors = {
  [key: string]: string
}

export type SiginInValues = {
  email: string
  password: string
}

export function signInValidate(values: SiginInValues): FieldErrors {
  const { email, password } = fieldsValidations
  const schema = Joi.object({ email, password })

  return getFieldErros(schema.validate(values, { abortEarly: false }))
}
