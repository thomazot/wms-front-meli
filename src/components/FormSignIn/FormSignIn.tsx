import { useState } from 'react'
import Router from 'next/router'

import Button from 'components/Button'
import TextField from 'components/TextField'

import { FieldErrors, signInValidate } from 'utils/validations'
import { FormWrapper } from 'styles/form'

export type FormSignInProps = {
  onSubmit?: () => void
}

const FormSignIn = ({ onSubmit }: FormSignInProps) => {
  const [inputs, setInputs] = useState({ email: '', password: '' })
  const [fieldError, setFieldError] = useState<FieldErrors>({})

  const handleInputs = (field: string, value: string) =>
    setInputs((inputsOld) => ({ ...inputsOld, [field]: value }))

  function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
    event.preventDefault()

    const errors = signInValidate(inputs)

    if (Object.keys(errors).length) {
      setFieldError(errors)
      return
    }
    setFieldError({})

    if (onSubmit) {
      onSubmit()
    }

    Router.push('/home')
  }

  return (
    <FormWrapper>
      <form onSubmit={handleSubmit}>
        <TextField
          name="email"
          placeholder="E-Mail"
          type="email"
          error={fieldError?.email}
          onInputChange={handleInputs}
        />
        <TextField
          name="password"
          placeholder="Password"
          error={fieldError?.password}
          type="password"
          onInputChange={handleInputs}
        />

        <Button type="submit" size="large" fullWidth>
          Sign in
        </Button>
      </form>
    </FormWrapper>
  )
}

export default FormSignIn
