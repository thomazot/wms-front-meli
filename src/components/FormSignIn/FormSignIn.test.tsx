import { render, screen } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import FormSignIn from '.'

jest.mock('next/router', () => ({ push: jest.fn() }))

describe('<FormSignIn />', () => {
  it('should render the form', () => {
    const { container } = render(<FormSignIn />)

    expect(screen.getByPlaceholderText(/e-mail/i)).toBeInTheDocument()
    expect(screen.getByPlaceholderText(/password/i)).toBeInTheDocument()

    expect(screen.getByRole('button', { name: /sign in/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should change values email and password', () => {
    render(<FormSignIn />)
    const email = screen.getByPlaceholderText(/e-mail/i)
    const password = screen.getByPlaceholderText(/password/i)

    userEvent.type(email, 'email@email.com')
    expect(email.getAttribute('value')).toBe('email@email.com')

    userEvent.type(password, 'password')
    expect(password.getAttribute('value')).toBe('password')

    const button = screen.getByRole('button', { name: /sign in/i })

    userEvent.click(button)
  })

  it('should submit form not called', () => {
    const onSubmit = jest.fn()
    render(<FormSignIn onSubmit={onSubmit} />)
    const button = screen.getByRole('button', { name: /sign in/i })

    userEvent.click(button)

    expect(onSubmit).not.toHaveBeenCalled()
  })

  it('should submit form called', () => {
    const onSubmit = jest.fn()
    render(<FormSignIn onSubmit={onSubmit} />)

    const button = screen.getByRole('button', { name: /sign in/i })

    const email = screen.getByPlaceholderText(/e-mail/i)
    const password = screen.getByPlaceholderText(/password/i)

    userEvent.type(email, 'email@email.com')
    userEvent.type(password, 'password')

    userEvent.click(button)

    expect(onSubmit).toHaveBeenCalled()
  })
})
