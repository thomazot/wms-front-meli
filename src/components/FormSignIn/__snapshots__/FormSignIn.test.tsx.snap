// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormSignIn /> should render the form 1`] = `
.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  color: #fff;
  background: #3483FA;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  border-radius: 0.4rem;
  cursor: pointer;
  font-size: 1.4rem;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
  width: 100%;
}

.c5:hover {
  background: #3270CC;
}

.c5:focus {
  background: #4988E8;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 0.4rem;
  padding: 0 1.6rem;
  border: 0.2rem solid #C5DCFF;
}

.c2:focus-within {
  box-shadow: 0 0 0.5rem #C5DCFF;
}

.c3 {
  font-size: 1.6rem;
  padding: 0.8rem 0;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c3:-webkit-autofill {
  -webkit-filter: none;
  filter: none;
}

.c3:-webkit-autofill::first-line {
  font-size: 1.6rem;
}

.c3:focus::-webkit-input-placeholder {
  color: transparent;
}

.c3:focus::-moz-placeholder {
  color: transparent;
}

.c3:focus:-ms-input-placeholder {
  color: transparent;
}

.c3:focus::placeholder {
  color: transparent;
}

.c0 .c1 {
  margin: 0.8rem 0;
}

.c0 .c4 {
  margin: 3.2rem auto 1.6rem;
}

<div
  class="c0"
>
  <form>
    <div
      class="c1 "
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          name="email"
          placeholder="E-Mail"
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="c1 "
    >
      <div
        class="c2"
      >
        <input
          class="c3"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="c4 c5"
      type="submit"
    >
      <span>
        Sign in
      </span>
    </button>
  </form>
</div>
`
